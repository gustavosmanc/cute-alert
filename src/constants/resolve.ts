export const ALERT_RESOLVE_VALUES = {
  PRIMARY_BUTTON_CLICKED: 'primaryButtonClicked',
  SECONDARY_BUTTON_CLICKED: 'secondaryButtonClicked',
  CLOSE: 'close'
} as const

export const TOAST_RESOLVE_VALUES = {
  CLOSE: 'close'
} as const
