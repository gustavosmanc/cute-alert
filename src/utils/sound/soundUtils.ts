export const playSound = (src?: string) => {
  if (src) {
    const sound = new Audio(src)
    sound.play()
  }
}
