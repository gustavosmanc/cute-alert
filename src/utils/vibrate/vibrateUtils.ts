export const vibrate = (pattern: number | number[]) => {
  navigator.vibrate(pattern)
}
