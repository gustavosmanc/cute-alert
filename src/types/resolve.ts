import { ALERT_RESOLVE_VALUES, TOAST_RESOLVE_VALUES } from '@/constants/resolve'

export type AlertResolveValue = (typeof ALERT_RESOLVE_VALUES)[keyof typeof ALERT_RESOLVE_VALUES]
export type ToastResolveValue = (typeof TOAST_RESOLVE_VALUES)[keyof typeof TOAST_RESOLVE_VALUES]

export type AlertResolve = (value: AlertResolveValue) => void
export type ToastResolve = (value: ToastResolveValue) => void
