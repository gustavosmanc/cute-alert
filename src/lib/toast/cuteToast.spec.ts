import { TOAST_RESOLVE_VALUES } from '@/constants/resolve'
import * as soundUtils from '@/utils/sound/soundUtils'
import * as vibrateUtils from '@/utils/vibrate/vibrateUtils'
import { queryAllByText, queryByText } from '@testing-library/dom'
import '@testing-library/jest-dom'
import { cuteToast } from '..'

const vibrateSpy = vi.spyOn(vibrateUtils, 'vibrate').mockImplementation(() => {})
const playSoundSpy = vi.spyOn(soundUtils, 'playSound').mockImplementation(() => {})

const getExampleDOM = () => {
  document.body = document.createElement('body')

  return document.body
}

const originalGlobalURL = global.URL

describe('cuteToast', () => {
  describe('snapshots', () => {
    beforeEach(() => {
      // @ts-expect-error
      global.URL = class URL {
        constructor(url: string) {
          // @ts-expect-error
          this.url = url
        }

        get href() {
          // @ts-expect-error
          return this.url
        }
      }
    })

    afterEach(() => {
      global.URL = originalGlobalURL
    })

    it('should render success toast', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'Title',
        description: 'Description'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render error toast', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'error',
        title: 'Title',
        description: 'Description'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render info toast', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'info',
        title: 'Title',
        description: 'Description'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render warning toast', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'warning',
        title: 'Title',
        description: 'Description'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render toast without title', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        description: 'Description'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render toast without description', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'Title'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render toast with timer', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'Title',
        description: 'Description',
        timer: 5000
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render two toasts', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'First Toast Title',
        description: 'First Toast Description'
      })

      cuteToast({
        type: 'success',
        title: 'Second Toast Title',
        description: 'Second Toast Description'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render toast with custom image size', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'Title',
        description: 'Description',
        imageSize: 48
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render toast with custom image source', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'Title',
        description: 'Description',
        imageSrc: 'random image source'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })
  })

  describe('functionalities', () => {
    it('should handle vibrate and playSound when toast is open', () => {
      getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'Title',
        description: 'Description',
        vibrationPattern: [1, 2, 3],
        soundSrc: 'sound source'
      })

      expect(vibrateSpy).toHaveBeenCalledTimes(1)
      expect(vibrateSpy).toHaveBeenCalledWith([1, 2, 3])
      expect(playSoundSpy).toHaveBeenCalledTimes(1)
      expect(playSoundSpy).toHaveBeenCalledWith('sound source')
    })

    it('should not close previous toast when opening a new one', () => {
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'First Toast Title',
        description: 'Description'
      })

      cuteToast({
        type: 'success',
        title: 'Second Toast Title',
        description: 'Description'
      })

      expect(queryByText(container, 'First Toast Title')).toBeInTheDocument()
      expect(queryByText(container, 'Second Toast Title')).toBeInTheDocument()
    })

    it('should handle close button click', async () => {
      const callbackMock = vi.fn()
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'Title',
        description: 'Description'
      }).then((response) => callbackMock(response))

      queryByText(container, 'X').click()

      await new Promise(process.nextTick)

      expect(container).toBeEmptyDOMElement()
      expect(callbackMock).toHaveBeenCalledTimes(1)
      expect(callbackMock).toHaveBeenCalledWith(TOAST_RESOLVE_VALUES.CLOSE)
    })

    it('should handle close button click with more than one toast', async () => {
      const callbackMock = vi.fn()
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'First Toast Title',
        description: 'Description'
      }).then((response) => callbackMock(response))

      cuteToast({
        type: 'success',
        title: 'Second Toast Title',
        description: 'Description'
      }).then((response) => callbackMock(response))

      queryAllByText(container, 'X')[0].click()

      await new Promise(process.nextTick)

      expect(queryByText(container, 'First Toast Title')).not.toBeInTheDocument()
      expect(queryByText(container, 'Second Toast Title')).toBeInTheDocument()
      expect(callbackMock).toHaveBeenCalledTimes(1)
      expect(callbackMock).toHaveBeenCalledWith(TOAST_RESOLVE_VALUES.CLOSE)
    })

    it('should handle close by timer', async () => {
      vi.useFakeTimers()

      const callbackMock = vi.fn()
      const container = getExampleDOM()

      cuteToast({
        type: 'success',
        title: 'Title',
        description: 'Description',
        timer: 5000
      }).then((response) => callbackMock(response))

      vi.advanceTimersByTime(4999)
      await new Promise(process.nextTick)

      expect(container).not.toBeEmptyDOMElement()
      expect(callbackMock).not.toHaveBeenCalled()

      vi.advanceTimersByTime(1)
      await new Promise(process.nextTick)

      expect(container).toBeEmptyDOMElement()
      expect(callbackMock).toHaveBeenCalledTimes(1)
      expect(callbackMock).toHaveBeenCalledWith(TOAST_RESOLVE_VALUES.CLOSE)

      vi.useRealTimers()
    })
  })
})
