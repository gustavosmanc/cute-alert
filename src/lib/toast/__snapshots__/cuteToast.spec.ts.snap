// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cuteToast > snapshots > should render error toast 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--error">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/error.svg" height="32" width="32">
          <div class="cute-toast__content">
            <span class="cute-toast__title">Title</span>
            <span class="cute-toast__description">Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div></div>"
`;

exports[`cuteToast > snapshots > should render info toast 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--info">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/info.svg" height="32" width="32">
          <div class="cute-toast__content">
            <span class="cute-toast__title">Title</span>
            <span class="cute-toast__description">Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div></div>"
`;

exports[`cuteToast > snapshots > should render success toast 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--success">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/success.svg" height="32" width="32">
          <div class="cute-toast__content">
            <span class="cute-toast__title">Title</span>
            <span class="cute-toast__description">Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div></div>"
`;

exports[`cuteToast > snapshots > should render toast with custom image size 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--success">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/success.svg" height="48" width="48">
          <div class="cute-toast__content">
            <span class="cute-toast__title">Title</span>
            <span class="cute-toast__description">Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div></div>"
`;

exports[`cuteToast > snapshots > should render toast with custom image source 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--success">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="random image source" height="32" width="32">
          <div class="cute-toast__content">
            <span class="cute-toast__title">Title</span>
            <span class="cute-toast__description">Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div></div>"
`;

exports[`cuteToast > snapshots > should render toast with timer 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--success">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/success.svg" height="32" width="32">
          <div class="cute-toast__content">
            <span class="cute-toast__title">Title</span>
            <span class="cute-toast__description">Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      <div class="cute-toast__timer" style="animation: timer 5000ms linear"></div>
    </div></div>"
`;

exports[`cuteToast > snapshots > should render toast without description 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--success">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/success.svg" height="32" width="32">
          <div class="cute-toast__content">
            <span class="cute-toast__title">Title</span>
            
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div></div>"
`;

exports[`cuteToast > snapshots > should render toast without title 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--success">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/success.svg" height="32" width="32">
          <div class="cute-toast__content">
            
            <span class="cute-toast__description">Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div></div>"
`;

exports[`cuteToast > snapshots > should render two toasts 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--success">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/success.svg" height="32" width="32">
          <div class="cute-toast__content">
            <span class="cute-toast__title">First Toast Title</span>
            <span class="cute-toast__description">First Toast Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div><div class="cute-toast cute-toast--success">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/success.svg" height="32" width="32">
          <div class="cute-toast__content">
            <span class="cute-toast__title">Second Toast Title</span>
            <span class="cute-toast__description">Second Toast Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div></div>"
`;

exports[`cuteToast > snapshots > should render warning toast 1`] = `
"<div class="cute-toast-wrapper"><div class="cute-toast cute-toast--warning">
      <div class="cute-toast__frame">
        <div class="cute-toast__body">
          <img class="cute-toast__image" src="/src/lib/icons/warning.svg" height="32" width="32">
          <div class="cute-toast__content">
            <span class="cute-toast__title">Title</span>
            <span class="cute-toast__description">Description</span>
          </div>
          <div class="cute-toast__close">X</div>
        </div>
      </div>
      
    </div></div>"
`;
